### PR TITLE
MAINT: io: replace isspmatrix with issparse in mmio module

### PR DESCRIPTION
**Pull request by dschult**
replace isspmatrix() with issparse() in mmio.py module

### CODE REVIEW

**Review by rossbar — APPROVED** [review]
LGTM - the sparse format writer is triggered by whether or no the input is sparse, so I believe `issparse` is the appropriate choice!

**Review by perimosocordiae — APPROVED** [review]

File: scipy/optimize/_nonlin.py [hunk 1]
```diff
@@ -480,7 +480,7 @@ def asjacobian(J):
                         solve=lambda v: solve(J, v),
                         rsolve=lambda v: solve(J.conj().T, v),
                         dtype=J.dtype, shape=J.shape)
-    elif scipy.sparse.isspmatrix(J):
+    elif scipy.sparse.issparse(J):
```
Comment (perimosocordiae) [inline scipy/optimize/_nonlin.py:483]: This should probably be reverted, at least for now: the code below is relying on `J*v` being a matrix-vector product, so a sparse array isn't a viable substitute here.

File: scipy/optimize/_nonlin.py [hunk 2]
```diff
@@ -516,7 +516,7 @@ def matvec(self, v):
                 m = J(self.x)
                 if isinstance(m, np.ndarray):
                     return dot(m, v)
-                elif scipy.sparse.isspmatrix(m):
+                elif scipy.sparse.issparse(m):
```
Comment (perimosocordiae) [inline scipy/optimize/_nonlin.py:519]: Same here. We want the sparse arrays to take the branch just above this one.

File: scipy/optimize/_nonlin.py [hunk 3]
```diff
@@ -534,7 +534,7 @@ def rmatvec(self, v):
                 m = J(self.x)
                 if isinstance(m, np.ndarray):
                     return dot(m.conj().T, v)
-                elif scipy.sparse.isspmatrix(m):
+                elif scipy.sparse.issparse(m):
```
Comment (perimosocordiae) [inline scipy/optimize/_nonlin.py:537]: And here, too.

**Review by perimosocordiae — CHANGES_REQUESTED** [review]

**Review by perimosocordiae — APPROVED** [review]
This is looking good now, thanks! Getting rid of that big block of dead code is quite nice.

**Review by rossbar — APPROVED** [review]
This LGTM as well! In principle the remaining `isspmatrix`es can/should also be updated to use `isinstance(J, spmatrix)` following #18531 , but that can be handled in another PR.

**Comment by dschult** [comment]
Yes, the `ispmatrix` calls in the optimize subpackage can be replaced by checking `isinstance(..., spmatrix)`. But it might be even better to adapt that code to get them working for either sparray or spmatrix (i.e. by changing from `*` to `@` and perhaps other changes).  So, I agree that changing the optimize code would be better in a separate PR as suggested.

And for future clarity, I'll explain here that the large block of code removed from '_mmio.py` in this PR is impossible to reach and thus no longer needed. Presumably it used to be able to be reached.  The import statement at the top of the module `from scipy.sparse import coo_matrix, ...` ensures that `coo_matrix` is present in scipy. So the local check at line 656 for `coo_matrix` will never result in `coo_matrix is None`.  That allows us to remove the lines for the if clause at line 719.

**Comment by dschult** [comment]
It'd be good to get this and #18565 merged before the new release so the deprecations in #18565 will be remove in v1.13.

**Comment by j-bowhay** [comment]
I believe the branch has already happened so these won't make it in until 1.12

**Review by rgommers — APPROVED** [review]
LGTM too. It seems to me like this is also a bug fix, for sparse array (rather than matrix) usage with this `io` functionality. So I'll mark it for backporting to 1.11.x

**Comment by rgommers** [comment]
Thanks @dschult, all.

**Pull request merged**